src/main/java/com/android/volley/CacheDispatcher.java

```diff
@@ -154,4 +180,92 @@ public void run() {
             }
         }
     }
+
+    /** Request received a valid response that can be used by other waiting requests. */
+    @Override
+    public void onResponseReceived(Request<?> request, Response<?> response) {
+        if (response.cacheEntry == null || response.cacheEntry.isExpired()) {
+            onNoUsableResponseReceived(request);
+            return;
+        }
+        String cacheKey = request.getCacheKey();
+        Queue<Request<?>> waitingRequests;
+        synchronized (mWaitingRequests) {
+            waitingRequests = mWaitingRequests.remove(cacheKey);
+        }
+        if (waitingRequests != null) {
+            if (VolleyLog.DEBUG) {
+                VolleyLog.v("Releasing %d waiting requests for cacheKey=%s.",
+                    waitingRequests.size(), cacheKey);
+            }
+            // Process all queued up requests.
+            for (Request<?> waiting : waitingRequests) {
+                mDelivery.postResponse(waiting, response);
+            }
+        }
+    }
+
+    /** No valid response received from network, release waiting requests. */
+    @Override
+    public void onNoUsableResponseReceived(Request<?> request) {
+        String cacheKey = request.getCacheKey();
+        Queue<Request<?>> waitingRequests;
+        synchronized (mWaitingRequests) {
+            waitingRequests = mWaitingRequests.remove(cacheKey);
+            if (waitingRequests != null) {
+                if (VolleyLog.DEBUG) {
+                    VolleyLog.v("%d waiting requests for cacheKey=%s; resend to network",
+                        waitingRequests.size(), cacheKey);
+                }
+                Request<?> nextInLine = waitingRequests.remove();
```

Comment:
How do you know nextInLine is not null? (A BlockingQueue does not accept null elements.)

Reply:
Added a check.